pkg/store/proxy.go

```diff
@@ -0,0 +1,242 @@
+	"github.com/improbable-eng/thanos/pkg/strutil"
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb/labels"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ProxyStore implements the store API that proxies request to all given underlying stores.
+type ProxyStore struct {
+	logger         log.Logger
+	stores         func() []*query.StoreInfo
+	selectorLabels labels.Labels
+}
+
+// NewProxyStore returns a new ProxyStore that uses the given clients that implements storeAPI to fan-in all series to the client.
+// Note that there is no deduplication support. Deduplication should be done on the highest level (just before PromQL)
+func NewProxyStore(
+	logger log.Logger,
+	stores func() []*query.StoreInfo,
+	selectorLabels labels.Labels,
+) *ProxyStore {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	s := &ProxyStore{
+		logger:         logger,
+		stores:         stores,
+		selectorLabels: selectorLabels,
+	}
+	return s
+}
+
+// Info returns store information about the external labels this store have.
+func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb.InfoResponse, error) {
+	res := &storepb.InfoResponse{
+		Labels: make([]storepb.Label, 0, len(s.selectorLabels)),
+	}
+	for _, l := range s.selectorLabels {
+		res.Labels = append(res.Labels, storepb.Label{
+			Name:  l.Name,
+			Value: l.Value,
+		})
+	}
+	return res, nil
+}
+
+// Series returns all series for a requested time range and label matcher. Requested series are taken from other
+// stores and proxied to RPC client. NOTE: Resulted data are not trimmed exactly to min and max time range.
+func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	match, newMatchers, err := labelsMatches(s.selectorLabels, r.Matchers)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	if !match {
+		return nil
+	}
+
+	var (
+		respCh    = make(chan *storepb.SeriesResponse, 10)
+		seriesSet []storepb.SeriesSet
+	)
+
+	for _, store := range s.stores() {
+		// We might be able to skip the store if its meta information indicates
+		// it cannot have series matching our query.
+		// NOTE: all matchers are validated in labelsMatches method so we explicitly ignore error.
+		if ok, _ := storeMatches(store, r.MinTime, r.MaxTime, newMatchers...); !ok {
+			continue
+		}
+		sc, err := store.Client.Series(srv.Context(), &storepb.SeriesRequest{
+			MinTime:  r.MinTime,
+			MaxTime:  r.MaxTime,
+			Matchers: newMatchers,
+		})
+		if err != nil {
+			respCh <- storepb.NewWarnSeriesResponse(errors.Wrap(err, "fetch series"))
+			continue
+		}
+
+		seriesSet = append(seriesSet, startStreamSeriesSet(sc, respCh, 10))
+	}
+
+	go func() {
+		mergedSet := storepb.MergeSeriesSets(seriesSet...)
+		for mergedSet.Next() {
+			var series storepb.Series
+			series.Labels, series.Chunks = mergedSet.At()
+			respCh <- storepb.NewSeriesResponse(&series)
+		}
```

Comment:
Error of mergedSet needs to be checked.

Reply:
damn, this will add additional channels ): Or... I could use errgroup here

Reply:
BTW stream is not thread safe https://groups.google.com/forum/#!topic/grpc-io/aI6L6M4fzQ0